crates/ruff_linter/src/rules/flake8_bandit/rules/hardcoded_sql_expression.rs

```diff
@@ -46,53 +45,77 @@ impl Violation for HardcodedSQLExpression {
     }
 }
 
-fn has_string_literal(expr: &Expr) -> bool {
-    string_literal(expr).is_some()
-}
-
-fn matches_sql_statement(string: &str) -> bool {
-    SQL_REGEX.is_match(string)
+/// Concatenates the contents of an f-string, without the prefix and quotes,
+/// and escapes any special characters.
+///
+/// ## Example
+///
+/// ```python
+/// "foo" f"bar {x}" "baz"
+/// ```
+///
+/// becomes `foobar {x}baz`.
+fn concatenated_f_string(expr: &ast::ExprFString, locator: &Locator) -> String {
+    expr.value
+        .parts()
+        .filter_map(|part| {
+            raw_contents(locator.slice(part)).map(|s| s.escape_default().to_string())
+        })
+        .collect()
 }
 
-fn matches_string_format_expression(expr: &Expr, semantic: &SemanticModel) -> bool {
-    match expr {
+/// S608
+pub(crate) fn hardcoded_sql_expression(checker: &mut Checker, expr: &Expr) {
+    let content = match expr {
         // "select * from table where val = " + "str" + ...
-        // "select * from table where val = %s" % ...
         Expr::BinOp(ast::ExprBinOp {
-            op: Operator::Add | Operator::Mod,
-            ..
+            op: Operator::Add, ..
         }) => {
             // Only evaluate the full BinOp, not the nested components.
-            if semantic
+            if !checker
+                .semantic()
                 .current_expression_parent()
                 .map_or(true, |parent| !parent.is_bin_op_expr())
             {
-                if any_over_expr(expr, &has_string_literal) {
-                    return true;
-                }
+                return;
+            }
+            if !any_over_expr(expr, &Expr::is_string_literal_expr) {
+                return;
             }
-            false
+            checker.generator().expr(expr)
+        }
+        // "select * from table where val = %s" % ...
+        Expr::BinOp(ast::ExprBinOp {
+            left,
+            op: Operator::Mod,
+            ..
+        }) => {
```

Comment:
This is now a separate branch because there could be implicit concatenations on the left side of `%` operator.